graphs/index.md

```diff
@@ -0,0 +1,706 @@
+allocation.
+
+### API Modifications
+
+```cpp
+namespace sycl {
+namespace ext::codeplay {
+
+// State of a queue, returned by info::queue::state
+enum class queue_state {
+  executing,
+  recording
+};
+
+// State of a graph
+enum class graph_state {
+  recordable,
+  executable
+};
+
+// New object representing graph
+template<graph_state State = graph_state::recordable>
+class command_graph {
+public:
+  /* Available only when: (State == graph_state::recordable) */
+  command_graph(const property_list &propList = {});
+  /* Available only when: (State == graph_state::recordable) */
+  command_graph<graph_state::executable> finalize(context &syclContext);
+  /* Available only when: (State == graph_state::executable) */
+  void update(const command_graph<graph_state::recordable> &graph);
+};
+}  // namespace ext::codeplay
+
+// New methods added to the sycl::queue class
+class queue {
+public:
+  bool begin_recording(command_graph<graph_state::recordable> &graph);
+  bool end_recording();
+  event submit(command_graph<graph_state::executable> graph);
+};
+}  // namespace sycl
+```
+
+#### Feature Test Macro
+
+This extension defines the [feature test macro][feature test macro]
+`SYCL_EXT_CODEPLAY_GRAPHS` which will be set to one of the following values
+depending on the version of the specification supported.
+
+| Value | Description            |
+|-------|------------------------|
+| 1     | First released version |
+
+#### New Graph Class
+
+This extension adds a new `command_graph` object which follows the
+[common reference semantics][common reference semantics] of other SYCL runtime
+objects.
+
+The `command_graph` object represents a collection of
+[command-group nodes](#node) and [their execution dependencies](#edge). A
+graph is built up by recording queue submissions, then once the user is happy
+that the graph is complete, the graph instance is finalized into an executable
+variant which can have no more nodes added to it. Finalization may be a
+computationally expensive operation as the runtime is able to perform
+optimizations based on the graph structure. After finalization the graph can be
+submitted for execution on a queue one or more times with reduced overhead.
+
+It is possible to submit an executable graph to a queue which is in the
+recording state. This will cause the graph being submitted to be added as a node
+in the graph being recorded to by the queue, this graph node is referred to as a
+*sub-graph*.
+
+##### Graph State
+
+An instance of a `command_graph` object can be in one of two states:
+
+* Recording - Graph is under construction and new nodes may be added to it.
+* Executable - Graph topology is fixed after finalization and graph is ready to
+  be submitted for execution.
+
+A `command_graph` object is constructed in the *recording* state and is made
+*executable* by the user invoking `command_graph::finalize()` to create a
+new executable instance of the graph. An *executable* graph cannot transition
```

Comment:
Transition could be misleading here. My understanding is, you keep the *editable* `command_graph` object alive, and create (multiple) *executable* object(s) from it, correct? Can `finalize()` method be `const`?

Reply:
> My understanding is, you keep the editable command_graph object alive, and create (multiple) executable object(s) from it, correct?

Yup, you understand correctly. *transition* does indeed sound like there is state change happening in-place, rather than creating something new. Will revisit this wording.

> `Can finalize()` method be const?

Not as we've currently implemented it, but conceptually I think this should be possible, and would make it clearer to the user about the point above of not modifying state in-place. Will take a look at refactoring that section of code to make this happen.

Reply:
We tweaked our implementation and indeed this can be `const`, updated to specify that